File: Python/Contains Duplicate II.py
class Solution:
    # @param {integer[]} nums
    # @param {integer} k
    # @return {boolean}
    def containsNearbyDuplicate(self, nums, k):
        length = min(k+1, len(nums))
        s = set()
        for i in range(0, length):
            if nums[i] in s:
                return True
            s.add(nums[i])
        if length == len(nums) + 1:
            return False
        for i in range(length, len(nums)):
            s.remove(nums[i-k-1])
            if nums[i] in s:
                return True
            s.add(nums[i])
        return False

File: java/Contains_Duplicate_II.java
// HashMap solution.
// Time is O(n), space is the hashmap size;

public class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        HashMap<Integer, Integer> table = new HashMap<Integer, Integer>();
        for (int i = 0; i < nums.length; ++i) {
            if (table.containsKey(nums[i])) {
                int j = table.get(nums[i]);
                if (i-j <= k) return true;
            }
            table.put(nums[i],i);
        }
        return false;
    }
}
